pip/pip-331.md

```diff
@@ -0,0 +1,129 @@
+    end
+    
+    develop --> runtime
+```
+
+# Detailed Design
+
+## Design & Implementation Details
+
+1. add `WasmLoader` to load WASM file and provide the WASM function to java, also provide the java function to WASM if we need.
+
+2. add `AbstractWasmFunction` and `AbstractWasmWindowFunction` as the core interface of the WASM function api.
+
+```java
+public abstract class AbstractWasmFunction<X, T> extends WasmLoader implements Function<X, T> {
+
+    private static final String PROCESS_METHOD_NAME = "process";
+
+    protected static final String INITIALIZE_METHOD_NAME = "initialize";
+
+    protected static final String CLOSE_METHOD_NAME = "close";
+
+    protected static final Map<Long, Argument<?>> ARGUMENTS = new ConcurrentHashMap<>();
+
+    @Override
+    public T process(X input, Context context) {
+        return super.getWasmExtern(PROCESS_METHOD_NAME)
+                .map(process -> {
+                    Long argumentId = callWASI(input, context, process);
+                    return doProcess(input, context, argumentId);
+                })
+                .orElseThrow(() -> new PulsarWasmException(
+                        PROCESS_METHOD_NAME + " function not found in " + super.getWasmName()));
+    }
+
+    private Long callWASI(X input,
+                          Context context,
+                          Extern process) {
+        // call WASI function
+        final Long argumentId = getArgumentId(input, context);
+        ARGUMENTS.put(argumentId, new Argument<>(input, context));
+        // WASI cannot easily pass Java objects like JNI, here we pass Long
+        // then we can get the argument by Long
+        WasmFunctions.consumer(super.getStore(), process.func(), WasmValType.I64)
+                .accept(argumentId);
+        ARGUMENTS.remove(argumentId);
+        return argumentId;
+    }
+
+    protected abstract T doProcess(X input, Context context, Long argumentId);
+
+    protected abstract Long getArgumentId(X input, Context context);
+
+    @Override
+    public void initialize(Context context) {
+        super.getWasmExtern(INITIALIZE_METHOD_NAME)
+                .ifPresent(initialize -> callWASI(null, context, initialize));
+    }
+
+    @Override
+    public void close() {
+        super.getWasmExtern(CLOSE_METHOD_NAME)
+                .ifPresent(close -> callWASI(null, null, close));
+        super.close();
+    }
+
+    protected static class Argument<X> {
+        protected X input;
+        protected Context context;
+
+        private Argument(X input, Context context) {
+            this.input = input;
+            this.context = context;
+        }
+    }
+}
+```
+
+More detailed code implementation and test can be found in [here](https://github.com/apache/pulsar/pull/21975)
+
+# Security Considerations
+
+Maybe need to add folders with tenancy name in the resource directory to prevent conflicts between WASM file names of different tenancies.
```

Comment:
What is the resource directory? Is it shared today by Pulsar functions?

Reply:
> What is the resource directory? 

For example, `resource/{tenancyName}/{wasmFileName}.wasm`

> Is it shared today by Pulsar functions?

It should not be shared, adding `tenancy name` is just to avoid path consistency conflicts.